internal/local/data.go

```diff
@@ -0,0 +1,176 @@
+
+	NameGraphQL         = "graphql"
+	NameCustomResolvers = "custom_resolvers"
+
+	NameHosting = "hosting"
+
+	NameSource = "source"
+
+	nameNodeModules = "node_modules"
+)
+
+// set of supported local files
+var (
+	FileRealmConfig = File{NameRealmConfig, extJSON}
+	FileConfig      = File{NameConfig, extJSON}
+	FileStitch      = File{NameStitch, extJSON}
+
+	FileSecrets = File{NameSecrets, extJSON}
+
+	FileSource = File{NameSource, extJS}
+)
+
+// File is a local Realm app file
+type File struct {
+	Name string
+	Ext  string
+}
+
+func (f File) String() string { return f.Name + f.Ext }
+
+// TODO(REALMC-7653): recursively walk the functions directory and collect all .js files
+// func walk(rootDir string, fn func(file os.FileInfo, path string) error) error {
+// 	dw := directoryWalker{path: rootDir}
+// 	if err := dw.walk(func(f os.FileInfo, p string) error {
+// 		if f.IsDir() {
+// 			return walk(p, fn)
+// 		}
+// 		return fn(f, p)
+// 	}); err != nil {
+// 		return err
+// 	}
+// 	return nil
+// }
+
+type directoryWalker struct {
+	path            string
+	continueOnError bool
+	failOnNotExist  bool
+	onlyDirs        bool
+	onlyFiles       bool
+}
+
+func (dw directoryWalker) walk(fn func(file os.FileInfo, path string) error) error {
+	if _, err := os.Stat(dw.path); err != nil {
+		if os.IsNotExist(err) && !dw.failOnNotExist {
+			return nil
+		}
+		return err
+	}
+	files, filesErr := ioutil.ReadDir(dw.path)
+	if filesErr != nil {
+		return filesErr
+	}
+	for _, file := range files {
+		if dw.onlyDirs && !file.IsDir() || dw.onlyFiles && file.IsDir() {
+			continue
+		}
+		err := fn(file, filepath.Join(dw.path, file.Name()))
+		if err != nil {
+			if dw.continueOnError {
+				continue
+			}
+			return err
+		}
+	}
+	return nil
+}
+
+func unmarshalDirectoryFlat(path string) ([]map[string]interface{}, error) {
+	var out []map[string]interface{}
+
+	dw := directoryWalker{path: path, onlyFiles: true}
+	if err := dw.walk(func(file os.FileInfo, path string) error {
+		switch filepath.Ext(path) {
+		case extJSON:
+			var o map[string]interface{}
+			if err := unmarshalJSONInto(&o, fileOptions{path: path}); err != nil {
+				return err
+			}
+			out = append(out, o)
+		}
+		return nil
+	}); err != nil {
+		return nil, nil
+	}
+
+	return out, nil
+}
+
+type fileOptions struct {
+	path          string
+	failOnEmpty   bool
+	failOnMissing bool
+}
+
+func defaultFileOptions(dir string, file File) fileOptions {
+	return fileOptions{path: filepath.Join(dir, file.String())}
+}
+
+func unmarshalJSONInto(out interface{}, opts fileOptions) error {
+	if _, err := os.Stat(opts.path); err != nil {
+		if os.IsNotExist(err) && !opts.failOnMissing {
+			return nil
+		}
+		return err
+	}
+
+	data, readErr := ioutil.ReadFile(opts.path)
+	if readErr != nil {
+		return fmt.Errorf("failed to read file at %s: %s", opts.path, readErr)
+	}
+
+	if len(data) == 0 {
+		if opts.failOnEmpty {
+			return fmt.Errorf("no file contents at %s", opts.path)
+		}
+		return nil
+	}
```

Comment:
what if we just make a helper func for this section? seems like all we need `out` for is to pass it directly to `json.Unmarshal`. I would suggest we create a helper function that returns `[]byte, error` from the file options. 

```
data, err := fileHelperBetterNamed(opts)
if err != nil { // handle }
err = json.Unmarshal(data, out)
if err != nil { // handle }
```

Maybe go one step further - we could just make the helper a default:
```
data, err := defaultFileHelperBetterNamed(path)
if err != nil { // handle }
err = json.Unmarshal(data, out)
if err != nil { // handle }
```

Reply:
yeah i like this, went for `unmarshalJSONFile` and `unmarshalJSONFileWithOptions`